Trim the README.md and add quickstart guide

niklas88 opened the pull request:
pretty aggressive trimming and a very quick no details no fuss quickstart guide

floriankramer commented on docs/advanced_features.md:
```diff
@@ -0,0 +1,71 @@
+
+QLever also supports efficient SPARQL autocompletion.  For example, the
+following query yields a list of all predicates associated with persons in the
+knowledge base, ordered by the number of persons which have that predicate.
```
`pesons` -> `people`

niklas88 replied:
Interestingly it turns out "persons" is [okay too](https://www.dailywritingtips.com/people-versus-persons/) and is in fact the more formal (older) form. Still, since we are using C++17 we might as well stick to the more modern form for English as well. So thanks for pointing this out to me!

floriankramer commented on docs/advanced_features.md:
```diff
@@ -0,0 +1,71 @@
+    GROUP BY ?predicate
+    ORDER BY DESC(?count)
+
+Note that this query could also be processed by standard SPARQL simply by
```
As SPARQL is the language, I would replace this with `Note that this query could also be processes by a standard SPARQL engine` or `Note that this query is equivalent to a standard SPARQL query`.

floriankramer commented on docs/advanced_features.md:
```diff
@@ -0,0 +1,71 @@
+    ORDER BY DESC(?count)
+
+Note that this query could also be processed by standard SPARQL simply by
+replacing the second triple by ?x ?predicate ?object. However, that query is
```
`replacying by` -> `replacuing with` that avoids having two repetitive bys. Also replace with is normally used in the active case (versus replace by in the passive case).

floriankramer commented on docs/advanced_features.md:
```diff
@@ -0,0 +1,71 @@
+
+Note that this query could also be processed by standard SPARQL simply by
+replacing the second triple by ?x ?predicate ?object. However, that query is
+bound to produce a very large intermediate result (all triples of all persons)
```
I would use `will` or `has to` instead of `is bound to`. That way, the sentence becomes simpler to read. Also `persons` -> `people`

floriankramer commented on docs/advanced_features.md:
```diff
@@ -0,0 +1,71 @@
+replacing the second triple by ?x ?predicate ?object. However, that query is
+bound to produce a very large intermediate result (all triples of all persons)
+with a correspondingly huge query time.  In contrast, the query above takes only
+~ 100ms on a standard Linux machine for a dataset with ~ 360 million triples and
```
This is down to personal preference, but I would write `about` instead of using `~` to interrupt the text flow less. Also, some rough information on the hardware of the linux machine could be interesting here, as I am otherwise uncertain what level of standard you refer to (standard for a consumer pc, or a compute server, ...)

niklas88 replied:
Hmm, the latter really depends on the KB. The example and the original text (written by Björn) is about Freebase Easy, which nowadays works well with 16 GB RAM so pretty much a beefy laptop. On the other hand we now think of Wikidata as the standard data set and that still needs 64 GB

niklas88 replied:
Since it says 360 million triples I think we should just mention 16 GB and it still works for "standard workstation"

floriankramer commented on docs/advanced_features.md:
```diff
@@ -0,0 +1,71 @@
+
+## Statistics:
+
+You can get stats for the currently active index in the following way:
```
While not wrong I suggest writing `statistics` instead of `stats` here. That sounds less colloquial to me.

floriankramer commented on docs/advanced_features.md:
```diff
@@ -0,0 +1,71 @@
+* The number of index permutations build (usually 2 or 6)
+* The numbers of distinct subjects, predicates and objects (only available if 6 permutations are built)
+* The name of the text index (if one is present)
+* The number of text records in the text index (if text index present)
```
To make the sentence structure more complete that should probably be `If a text index is present`

floriankramer commented on docs/advanced_features.md:
```diff
@@ -0,0 +1,71 @@
+* The number of entity occurrences/postings in the text index (if text index present)
+
+
+The names of a index is the name of the input nt file (and wordsfile for the text index) but can also be specified manually while building an index.
```
`The names` -> `The name` and `a` -> `an`. I believe it should also be `nt input file` but I'm uncertain if there is any rule to back up that feeling. `while` -> `when` as specifying the name is more of a single action and not a process.
If an index can have two names (a text and nt based name) the `names` is fine of course, but you then need to use `of a index are` instead of `is`

floriankramer commented on docs/advanced_features.md:
```diff
@@ -0,0 +1,71 @@
+
+Currently, QLever does not compute partial results if there is a LIMIT modifier.
+However, strings (for entities and text excerpts) are only resolved for those items that are actually send.
+Furthermore, in a UI, it is usually beneficial to get less than all result rows by default.
```
I would restructure this sentence to avoid the `less than all`, as it sounds akward to me. For example:
`Furthermore, a UI usually only [requires/wants] a limited amount of rows at a time.`

floriankramer commented on docs/advanced_features.md:
```diff
@@ -0,0 +1,71 @@
+However, strings (for entities and text excerpts) are only resolved for those items that are actually send.
+Furthermore, in a UI, it is usually beneficial to get less than all result rows by default.
+
+While it is recommended for applications to specify a LIMIT, some experiments want to measure the time to produce the full result but not block the UI.
```
To me `While specifying a limit is recommended, some experiments may want to` has a better flow.

floriankramer commented on docs/advanced_features.md:
```diff
@@ -0,0 +1,71 @@
+Furthermore, in a UI, it is usually beneficial to get less than all result rows by default.
+
+While it is recommended for applications to specify a LIMIT, some experiments want to measure the time to produce the full result but not block the UI.
+Therefore an additional HTTP parameter "&send=<x>" can be used to only send x result rows but to compute the fully readable result for everything according to LIMIT.
```
`to only send` -> `to send only` `but to compute` -> `while still computing the readable result for up to LIMIT rows`

floriankramer commented on docs/advanced_features.md:
```diff
@@ -0,0 +1,71 @@
+While it is recommended for applications to specify a LIMIT, some experiments want to measure the time to produce the full result but not block the UI.
+Therefore an additional HTTP parameter "&send=<x>" can be used to only send x result rows but to compute the fully readable result for everything according to LIMIT.
+
+**IMPORTANT: Unless you want to measure QLever's performance, using LIMIT (+ OFFSET for sequential loading) should be preferred in all applications. That way should be faster and standard SPARQL without downsides.**
```
`should be` -> `is`
`That way...` -> `LIMIT is faster and produces the same output as the send parameter.`

floriankramer commented on docs/quickstart.md:
```diff
@@ -0,0 +1,47 @@
+# Quickstart
+The following instructions build an index from our scientists example Knowledge
+Base
```
`Base` -> `Base.`

floriankramer commented on docs/quickstart.md:
```diff
@@ -0,0 +1,47 @@
+        --name qlever \
+        qlever
+
+Then point your browser to [http://localhost:7001/](http://localhost:7001/) and
```
I would just write `Then open ... in your browser.`

floriankramer commented on docs/quickstart.md:
```diff
@@ -0,0 +1,47 @@
+enter the query.
+
+For example all female scientists occuring in the text with "algo*" can be
+obtained with the following
```
`in the text` -> `in the text corpus` as no text was mentioned right before this sentence. `with "algo*"` -> `with the regex "algo*"`
`with the following` -> `with the folllowing query`

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+# Troubleshooting
+
+If you have problems, try to rebuild when compiling with `-DCMAKE_BUILD_TYPE=Debug`.
```
-> `If you have problems, try rebuilding QLever in debug mode (`cmake . -DCMAKE_BUILD_TYPE=Debug -DLOGLEVEL=DEBUG && make -j4`)`

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+# Troubleshooting
+
+If you have problems, try to rebuild when compiling with `-DCMAKE_BUILD_TYPE=Debug`.
+In particular also rebuild the index.
```
-> `Then rebuild your index with the new IndexBuilderMain executable.`

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+## Run End-to-End Tests
+
+QLever includes a simple mechanism for testing the entire system from
+from building an index to running queries in a single command.
```
`from from` -> `from`
I awould also omit the `in a single command.` as it complicates the sentence structure and we have already claimed the mechanism to be simple.

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+QLever includes a simple mechanism for testing the entire system from
+from building an index to running queries in a single command.
+
+In fact this End-to-End Test is run on Travis CI for every commit and Pull
```
`In fact this` -> `In fact, the`

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+
+In fact this End-to-End Test is run on Travis CI for every commit and Pull
+Request. It is also useful for local development however since it allows you to
+quickly test if something is horribly broken.
```
`development however since` -> `development, as`
Do we need to advertise the general benefits of good testing here?  This seems like a strange sentence in a troubleshooting documents for people trying out qlever.

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+quickly test if something is horribly broken.
+
+Just like QLever itself the End-to-End Tests can be executed from a previously
+build QLever container
```
`contiainer` -> `docker container` (I assume that is what you meant here)

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+
+**Note**: If you encounter permission issues e.g. if your UID is not 1000 or you
+are using docker with user namespaces, add the flag `-u root` to your `docker
+run` command or do a `chmod -R o+rw e2e_data`
```
The same needs to be done for the `src/web` folder (as far as I know) as docker will otherwise potentially be unable to copy the index.html.

niklas88 replied:
The copying is done by the `docker` daemon which runs as root. The access is only a problem when the repository is checked out on an NFS where `root` can't read files owned by the normal user. That is a pretty obscure problem, though sadly it happens more often on our infrastructure as our homes are configured that way.

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+    docker build -t qlever .
+    docker run -it --rm -v "$(pwd)/e2e_data:/app/e2e_data/" --name qlever-e2e --entrypoint e2e/e2e.sh qlever
+
+## Converting old Indices For Current QLever Versions
```
`old` should also be capitalized, as `Current` is capitalized.

niklas88 replied:
I've made "current" lower case. Funnily enough the German "upper case nouns and the start of a sentence" algorithm is a valid title casing in English and actually the one I was aiming for

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+
+## Converting old Indices For Current QLever Versions
+
+We have recently updated the way the index meta data (offsets of relations
```
Dates or commit hashes would be useful here.

niklas88 replied:
I think this will too easily get out of date. The idea is that we support all breaking index changes in a similar procedure and detecting the need for it at load time. There are index changes which only unbreak things (e.g. the current PR for `xsd:double`) but keep all functionality.

I've reworded to reflect this

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+will work but throw a warning at runtime.
+
+We have provided a converter which
+allows to only modify the meta data without having to rebuild the index. Just
```
Remove `allows to`, `modify` -> `modifies`

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+in order to use the converted index (rename to `<index-prefix>.index.ops` in our
+example).
+**Please consider creating backups of
+the "original" index files before overwriting them like this**.
```
`like this` -> ``

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+
+## High RAM Usage During Runtime
+
+QLever uses an on-disk index and is usually able to operate with pretty low RAM
```
`.. and is usually...` -> `and usualy requires little RAM.`. `pretty` is colloquial in this context.

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+## High RAM Usage During Runtime
+
+QLever uses an on-disk index and is usually able to operate with pretty low RAM
+usage. For example it can handle the full Wikidata KB + Wikipedia which is about
```
`For example` -> `For example,` `Wikipedia` -> `Wikipedia, `

floriankramer commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,73 @@
+1.5 TB of index with less than 46 GB of RAM
+
+However, there are data layouts that currently lead to an excessive
+amount of memory being used.
```
Without more context this sentence is confusing. Which data layouts are the problem? Is there something a user should avoid doing?
If this references the list below, you should use a colon instead of a full stop at the end of the sentence.

floriankramer commented on docs/wikidata.md:
```diff
@@ -0,0 +1,92 @@
+# Wikidata Full Quickstart
+
+The following instructions take you through loading the full Wikidata Knowledge
+Base in QLever. Since Wikidata is very large at over 7 billion triples, **we
```
`in` -> either `into` or `using`

floriankramer commented on docs/wikidata.md:
```diff
@@ -0,0 +1,92 @@
+The following instructions take you through loading the full Wikidata Knowledge
+Base in QLever. Since Wikidata is very large at over 7 billion triples, **we
+recommend to run all of the following commands in a `tmux` or `screen` session**.
+Nevertheless we will note below when a command will take long to execute.
```
`` -> `Commands that take a long time will be marked as such.`
The current sentence implies that the reader will ignore the advice we gave him, which could be considered insulting.

floriankramer commented on docs/wikidata.md:
```diff
@@ -0,0 +1,92 @@
+Nevertheless we will note below when a command will take long to execute.
+
+## Download and build QLever using `docker`
+
```
Add a link to docker, or short instructions on how to install it for ubuntu.

floriankramer commented on docs/wikidata.md:
```diff
@@ -0,0 +1,92 @@
+If you already downloaded **and decrompressed** Wikidata to uncompressed Turtle
+format you can skip this step, otherwise we download and uncompress it.
+
+**Note:** This takes several hours as Wikidata is about 42 GB compressed and the
```
`the` -> `their`

floriankramer commented on docs/wikidata.md:
```diff
@@ -0,0 +1,92 @@
+
+**Note:** This takes several hours as Wikidata is about 42 GB compressed and the
+servers are throttled. Also make sure you have enough space on the filesystem
+where you unpacked QLever or see the full
```
`unpacked` -> `unpack` (we haven't unpacked QLever yet)

niklas88 replied:
Good point, I was thinking about downloading a `tar` of QLever itself even though the instructions do a `git clone`

floriankramer commented on docs/wikidata.md:
```diff
@@ -0,0 +1,92 @@
+[README](https://github.com/ad-freiburg/QLever#building-the-index) for
+instructions on using a different path for the index.
+
+**The index plus unpacked Wikidata will use up to about 2 TB.**
```
`plus` -> `together with the`,  `Wikidata` -> `Wikidata turtle file`, `up to about` -> `about`, unless there is a case where it uses significantly less.

floriankramer commented on docs/wikidata.md:
```diff
@@ -0,0 +1,92 @@
+## Build a QLever Index
+
+Now we can build a QLever Index from the `latest-all.ttl` Wikidata Turtle file
+using the `wikidata_settings.json` file for some useful default settings for
```
`... .json file for some ...`  this sentence is ginourmous and should probably be split into two. E.g.:
`.json file. The .json file conatains some useful...`

floriankramer commented on docs/wikidata.md:
```diff
@@ -0,0 +1,92 @@
+
+Now we can build a QLever Index from the `latest-all.ttl` Wikidata Turtle file
+using the `wikidata_settings.json` file for some useful default settings for
+relations that can be safely stored on disk because their actual values are
```
`that can be stored on disk safely as their`

floriankramer commented on docs/wikidata.md:
```diff
@@ -0,0 +1,92 @@
+-u`) is not 1000 you have to make the `./index` folder writable for QLever
+inside the container e.g. by running `chmod -R o+rw ./index`
+
+**Note (1):** This takes about half a day but should be much faster than with most
```
`but should be` ->`which is still faster than [most] other triple stores.`. I would consider simply removing that part of the sentence though, as this is a tutorial and not a comparison, and it feels somewhat out of place to me.

floriankramer commented on docs/wikidata.md:
```diff
@@ -0,0 +1,92 @@
+        qlever
+
+Then point your browser to [http://localhost:7001/](http://localhost:7001/) and
+enter the query.
```
Same as before, I prefer `open ... in your browser` and I would remove the `enter the query` part, as that should be obvious, and the implication that the user wants to enter exactly one query sounds strange to me.

floriankramer commented on docs/wikidata.md:
```diff
@@ -0,0 +1,92 @@
+Then point your browser to [http://localhost:7001/](http://localhost:7001/) and
+enter the query.
+
+For example the following query retrieves all mountains above 8000 m
```
`For example, `

niklas88 commented:
@floriankramer thanks for the review. I think I addressed all of your comments. I also reworked the Wikidata Quickstart Guide a bit

* Moved the filesystem space part to the `git clone` so it's less likely one needs to move it later
* Changed to using a `wikidata-input` folder so it's more clear how the input can be stored separately
* Uncompress wikidata directly while downloading, so that it is one slow command and one can get started immediately on the next steps once that is finished

niklas88 commented on README.md:
```diff
@@ -3,106 +3,101 @@
+Alternatively to get started with a real (and really big) dataset we have prepared
+a [Wikidata Quickstart Guide](docs/wikidata.md). This guide takes you through the entire
+process of loading the full Wikidata Knowledge Base into QLever, but don't worry
+it is pretty.
```
I guess I missed a flush in my brain's output path.

niklas88 commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,77 @@
+When we make changes to the way the index meta data (e.g. offsets of relations
+within the permutations) is stored. Old index builds with 6 permutations will
+not work directly with newer QLever version while 2 permutation indices
+will work but throw a warning at runtime. This will be detected during startup.
```
It also detects when it won't work at all in which case it refuses to start. I will just remove the "but throw a warning" part

niklas88 commented on docs/troubleshooting.md:
```diff
@@ -0,0 +1,77 @@
+causes long and rarely used strings to be externalized to
+disk. This saves a significant amount of memory at little to no time cost for
+typical queries. The strategy can be modified to be more aggressive (currently
+by editing directly in the code during index construction)
```
This is actually addressed by my other pending PR #207 so I will just remove this comment. I expect we will merge the README changes before that PR and then I can add documentation for the settings there.

niklas88 commented:
@floriankramer thank you for the great (as always) review. I've addressed your comments. I'm really looking forward to the new README. I also used the commands in the Wikidata Quickstart Guide for building the `xsd:double` index yesterday, so these definitely work.

floriankramer commented on README.md:
```diff
@@ -55,8 +55,9 @@ Further documentation is available on the following topics
 
-We recommend using QLever with `docker` if you absolutely want to run QLever
-directly on your host see [here](docs/native_setup.md).
+We recommend using QLever with [docker](https://www.docker.com) if you
```
`if` -? `. If`

floriankramer commented on docs/advanced_features.md:
```diff
@@ -30,11 +30,13 @@ knowledge base, ordered by the number of people which have that predicate.
-a correspondingly long query time.  In contrast, the query above takes only
-about 100 ms on a standard Linux machine (with 16 GB memory) and a dataset with 360
-million triples and 530 million text records.
+by replacing the second triple with `?x ?predicate ?object` and add `DISTINCT`
```
`add` -> `adding`

floriankramer commented on docs/wikidata.md:
```diff
@@ -27,12 +29,14 @@ build the index under a different path.
 
 If you already downloaded **and decrompressed** Wikidata to uncompressed Turtle
-format you can skip this step, otherwise we download and uncompress it.
+format you can skip this step. Otherwise we download and uncompress it.
```
`to uncompressed` -> `to the uncompressed`
I'm probably just nitpicking at this point, but the `Otherwise we download and uncompress it.` still sounds slightly weird to me. What about `Otherwise we'll download and uncompress it in this step.`
